xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -1429,21 +1442,75 @@ public void handleRdsResponse(
         Collections.<String>emptySet(), versionInfo, nonce, errors);
   }
 
-  private static RdsUpdate processRouteConfiguration(
+  private static List<VirtualHost> processRouteConfiguration(
       RouteConfiguration routeConfig, FilterRegistry filterRegistry, boolean parseHttpFilter)
       throws ResourceInvalidException {
+    Map<String, PluginConfig> pluginConfigMap = new HashMap<>();
+    if (enableRouteLookup) {
+      List<ClusterSpecifierPlugin> plugins = routeConfig.getClusterSpecifierPluginsList();
+      for (ClusterSpecifierPlugin plugin : plugins) {
+        PluginConfig existing = pluginConfigMap.put(
+            plugin.getExtension().getName(), parseClusterSpecifierPlugin(plugin));
+        if (existing != null) {
+          throw new ResourceInvalidException(
+              "Multiple ClusterSpecifierPlugins with the same name: "
+                  + plugin.getExtension().getName());
+        }
+      }
+    }
     List<VirtualHost> virtualHosts = new ArrayList<>(routeConfig.getVirtualHostsCount());
     for (io.envoyproxy.envoy.config.route.v3.VirtualHost virtualHostProto
         : routeConfig.getVirtualHostsList()) {
       StructOrError<VirtualHost> virtualHost =
-          parseVirtualHost(virtualHostProto, filterRegistry, parseHttpFilter);
+          parseVirtualHost(virtualHostProto, filterRegistry, parseHttpFilter, pluginConfigMap);
       if (virtualHost.getErrorDetail() != null) {
         throw new ResourceInvalidException(
             "RouteConfiguration contains invalid virtual host: " + virtualHost.getErrorDetail());
       }
       virtualHosts.add(virtualHost.getStruct());
     }
-    return new RdsUpdate(virtualHosts);
+    return virtualHosts;
+  }
+
+  private static PluginConfig parseClusterSpecifierPlugin(ClusterSpecifierPlugin pluginProto)
+      throws ResourceInvalidException {
+    return parseClusterSpecifierPlugin(
+        pluginProto, ClusterSpecifierPluginRegistry.getDefaultRegistry());
+  }
+
+  @VisibleForTesting
+  static PluginConfig parseClusterSpecifierPlugin(
+      ClusterSpecifierPlugin pluginProto, ClusterSpecifierPluginRegistry registry)
+      throws ResourceInvalidException {
+    TypedExtensionConfig extension = pluginProto.getExtension();
+    String pluginName = extension.getName();
+    Any anyConfig = extension.getTypedConfig();
+    String typeUrl = anyConfig.getTypeUrl();
+    Message rawConfig = anyConfig;
+    try {
+      if (typeUrl.equals(TYPE_URL_TYPED_STRUCT_UDPA)) {
+        TypedStruct typedStruct = anyConfig.unpack(TypedStruct.class);
+        typeUrl = typedStruct.getTypeUrl();
+        rawConfig = typedStruct.getValue();
+      } else if (typeUrl.equals(TYPE_URL_TYPED_STRUCT)) {
+        com.github.xds.type.v3.TypedStruct newTypedStruct =
+            anyConfig.unpack(com.github.xds.type.v3.TypedStruct.class);
+        typeUrl = newTypedStruct.getTypeUrl();
+        rawConfig = newTypedStruct.getValue();
```

Comment:
I think this quote implies xds and udp TypedStruct types are compatible:

> In the case that the type URL is xds.type.v3.TypedStruct (or, for historical reasons, udpa.type.v1.TypedStruct), the inner type URL of TypedStruct will be utilized.  
> — https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/extension.proto#config-core-v3-typedextensionconfig

Then we can simplify this code by using unpackCompatibleType().
```java
TypedStruct typedStruct = unpackCompatibleType(anyConfig, TypedStruct.class, TYPE_URL_TYPED_STRUCT_UDPA, TYPE_URL_TYPED_STRUCT);
```

Reply:
Done.